teps/0014-step-timeout.md

```diff
@@ -0,0 +1,174 @@
+---
+title: step-timeout
+authors:
+  - "@Peaorl"
+creation-date: 2020-09-10
+last-updated: 2020-09-10
+status: proposed
+---
+
+# TEP-0014: Step timeout
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal<a name="Proposal"></a>](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Caveats](#caveats)
+    - [Resolution of a <code>Step</code> Timeout](#resolution-of-a--timeout)
+- [Test Plan](#test-plan)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
```

Comment:
What does the spec look for the taskRun? I think the smallest change might be introducing this at the run level first.    If I use another person's task, Do I have to manually edit their task?

Reply:
This might be a dumb question, but this seems unclear to me.

Reply:
I think step-level details belong as a configuration of the Task, and should in general be opaque to TaskRun users. This lets a Task author change the implementation of their Task without requiring any change from users running that Task (hopefully 🤞).

If a Task author changes a step's timeout in a way that the TaskRun-er isn't comfortable with, they should fork and redefine the Task, or inline the TaskSpec definition in their TaskRun.

Reply:
Indeed in that case you would have to edit the person's `Task`. The `TaskRun` spec doesn't change.

The reason for adding it to the `Task` and not the `TaskRun` is that it's assumed the `Task` author will have the most knowledge about how  the `Step` should behave.

Also latching on to [this](https://github.com/tektoncd/pipeline/pull/510#issuecomment-468378731) comment, it's suggested `TaskRuns` should be used for variations between invocations. I think in this case usually the timeout specification will remain the same between `TaskRun`/`PipelineRun` invocations.

Reply:
cool. That makes sense. It would also be pretty neat if as a task author, I could parameterize this field and give it a default value.

Reply:
I believe that substitution is done at run-state. So, if timeout field is typed for non-string. Im not sure if that would be possible. 🤔  (But, I might be getting to much in the weeds here) Anyway, this lgtm. 😄

Reply:
Indeed that happens at run-time. 

In what scenario would you like to see this parametrized? Would that be to reuse a timeout specification across `Steps`.
Because I would say that every `Step` (that is desirable to timeout) should be scrutinized in order to find a fitting timeout for the `Step`. Therefore I'm not sure a timeout is something you would want to pass as a parameter in a `Task`.

Reply:
I have task that is a `bash script runner`. I provide it a git repo with a bash script and it runs it. I would like to decide how long the timeout is for that step. I don't want to use a task timeout because the first step is a git-clone. I would like to set a timeout on the script run step depending on what was pulled.